### DISCUSSION TITLE
Exposing simulator info to CLI/GUI or other external tools

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#1065] [Ideas]
Hello Shadow community!

Here I explain an idea for extending Shadow that we are considering. My hope is to gauge the level of interest among those who use or are interested in Shadow.

The idea is analogous to what [Tor](https://www.torproject.org/) does with it's control port. Tor has a [specification](https://gitweb.torproject.org/torspec.git/tree/control-spec.txt) that describes the protocol for querying information from Tor and what types of information is available. They also have client libraries like [stem](https://stem.torproject.org/) and user interfaces like [nyx](https://nyx.torproject.org/) that are built on top of the control API. The idea is to do something similar to Shadow.

The idea is to develop an API through which external tools could query a running Shadow during a simulation for various information about the ongoing simulation. This would enable tools to more easily discover the state and progress of a long-running simulation, and would allow external CLI/GUI clients to be built to help users visualize a simulation while it is running. If we had such an API in shadow, we could ask contributors or other interested volunteers to consider working on a CLI/GUI that makes use of the information we expose. We could start by exposing basic information and extend the API over time.

Here is an incomplete list of basic information that we might consider making available through such an API:
- current simulation time and run time
- the number of events that have already been processed (global, per host)
- the number of outstanding events that need to be processed (global, per host)
- names and IP addresses of all hosts running in the simulation
- number of plugins (processes) that a given host is running
- number of open sockets, files, or other descriptors for a given host

Questions for discussion:
- Is this API something anyone would find useful?
- Is would anyone be interested in working on a CLI/GUI client that would make use of this API if we were to develop it?
- Is there some useful piece of information not included above that should be added to the list?

**Comment by jtracey** [comment]
Some other metrics that would be useful for spotting Shadow performance bottlenecks would be
 - execution time spent on a given host
 - thread idle time

The latter is already recorded, I don't remember if the former is.

Something that seems useful generally would be a way to extend this API from the plugins. That way you could do things like get information from Tor's control port via Shadow's API (e.g., monitor whether the network is bootstrapping correctly).
